### PR TITLE
Push dialogue refactor

### PR DESCRIPTION
**Pull request by zmerr**
It seemed that a lot of the code in GitPushDialog was identical to the one existing in VcsPushDialog.
This PR removes this redundant code.

### CODE REVIEW

**Comment by mkondratek** [comment]
oh, are we sure about it? I do remember I did some hackish workaround there but I do not reflect the reason 🤔 lemme take a deeper look into it. I hope you are right, and it is no longer needed, and we can drop it

**Comment by mkondratek** [comment]
ah, I see the reason... look

image

IIRC the whole workaround here was to have `Force Push` instead of `Push` button in the dialog in the case when we know it is required 😞 I am afraid we cannot drop it

**Comment by zmerr** [comment]
> ah, I see the reason... look
> 
> image
> 
> IIRC the whole workaround here was to have `Force Push` instead of `Push` button in the dialog in the case when we know it is required 😞 I am afraid we cannot drop it

Well, I would say, maybe we can just rewrite the constructor instead of calling the `super` one. WDYT?

**Comment by zmerr** [comment]
@mkondratek I tested the functionality with the latest IntelliJ version. As you mentioned, the force push does not appear anymore. But there is something even better. It suggests to rebase on the remote or to merge remote into local, and it does it correctly.
I mean when you select rebase, it interactively asks you whether to accept your version or theirs. So WDYT?

**Comment by mkondratek** [comment]
> @mkondratek I tested the functionality with the latest IntelliJ version. As you mentioned, the force push does not appear anymore. But there is something even better. It suggests to rebase on the remote or to merge remote into local, and it does it correctly. I mean when you select rebase, it interactively asks you whether to accept your version or theirs. So WDYT?

I see two problems :no_entry: :
1. The dialog that suggests the rebase on the remote or the merge is not clear. I am not sure what is to be rebased on what (or merged into what).
image

2. We cannot drop the default `Force Push` option in the push dialog for this one. It’s a primary method of resolution to the divergency b/w the local and the remote. As we have discussed it somewhere before, the rebase on the remote and the merge of the remote into the local are additional options that we very likely will consider and add to the plugin, but for now there are other priorities.

However, I see one hope too :star: :
1. Probably, when the time comes we will be able to reuse the code more effectively as the mentioned features are already implemented by git4idea.

 ---

 In the scope of these changes, I will write it again - we cannot drop the `Force Push`-by-default functionality :slightly_frowning_face:

**Comment by zmerr** [comment]
@mkondratek Thank you so much for the input, Mikolaj. I think since there is no straightforward way to modify the push button text, we need to keep the code of the GitPushDialog as it is, for now.

As for the traverse feature, since pushing on remote appears to be non-interfering in other tasks of traversing, maybe we can just initiate the `GitPushDialog` and immediately after calling `.show` on that proceed on processing the next branch, allowing the push functionality to occur in the background for itself. WDYT?

> Probably, when the time comes we will be able to reuse the code more effectively as the mentioned features are already implemented by git4idea.

Do you mean the code for rebase and merge on remote which IntelliJ already has? Maybe later by modification of the dialog to make the message more clear or by creating a ticket on the IntelliJ itself.

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GitPushDialog.java [hunk 1]
```diff
@@ -1,339 +1,26 @@
...
-  }
-
-  @UIEffect
-  private void push() {
-    push(isForcePushRequired);
-  }
-
-  @Override
-  @UIEffect
-  public void executeAfterRunningPrePushHandlers(Task.Backgroundable activity) {
```
[the remark this answers is not in the document]
Comment (zmerr) [reply frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GitPushDialog.java:238]: maybe we can just drop the `@Override`

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GitPushDialog.java [hunk 2]
```diff
@@ -1,339 +1,26 @@
...
-      close(OK_EXIT_CODE);
-    } else if (result == PrePushHandler.Result.ABORT_AND_CLOSE) {
-      doCancelAction();
-    } else if (result == PrePushHandler.Result.ABORT) {
-      // cancel push and leave the push dialog open
-    }
-  }
-
-  @UIEffect
-  public PrePushHandler.Result runPrePushHandlersInModalTask() {
```
Comment (zmerr) [inline frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GitPushDialog.java:251]: this method seems to be identical to the one in the `VcsPushDialog` too.

**Comment by mkondratek** [comment]
> @mkondratek Thank you so much for the input, Mikolaj. I think since there is no straightforward way to modify the push button text, we need to keep the code of the GitPushDialog as it is, for now.
> 
> As for the traverse feature, since pushing on remote appears to be non-interfering in other tasks of traversing, maybe we can just initiate the `GitPushDialog` and immediately after calling `.show` on that, proceed on processing the next branch, allowing the push functionality to occur in the background for itself. WDYT?
> 
> > Probably, when the time comes we will be able to reuse the code more effectively as the mentioned features are already implemented by git4idea.
> 
> Do you mean the code for rebase on remote, or merge remote into local which IntelliJ already has? Maybe later by modification of the dialog to make the message more clear or by creating a ticket on IntelliJ itself.

Yes! ✨  As you can see in https://github.com/VirtusLab/git-machete-intellij-plugin/pull/1352 we can use `GitPushDialog` within the traverse feature without any refactor needed at this point 🙅  Actually I do not recall much changes on that while I was taking your changes - it seems that you had implemented it already correctly (or almost correctly) 🙂  anyway works now 🚀 

> > In the scope of these changes, I will write it again - we cannot drop the Force Push-by-default functionality 🙁
> 
> The IntelliJ’s own dialogue initiated from the git toolbar, (not git machete), does have the force push button, but in VcsPushDialog it is disabled by a default false flag for force push. Since some fields are private, they can’t be overridden or modified.

Sadly, yes 🙁 the access to the field is restricted and that must be the reason why I had implemented it this way in the first place 😞

**Pull request closed without merging**